app/models/supplejack_api/top_collection_metric.rb

```diff
@@ -0,0 +1,136 @@
+# frozen_string_literal: true
+
+module SupplejackApi
+  # app/models/supplejack_api/top_collection_metric.rb
+  class TopCollectionMetric
+    include Mongoid::Document
+
+    METRICS_LOGGER = Logger.new('log/metrics.log') # run `$ touch log/metrics.log` to create metrics log file
+
+    METRICS = %i[
+      page_views
+      user_set_views
+      user_story_views
+      added_to_user_sets
+      source_clickthroughs
+      appeared_in_searches
+      added_to_user_stories
+    ].freeze
+
+    field :d, as: :date,               type: Date, default: Time.current.utc
+    field :m, as: :metric,             type: String
+    field :r, as: :results,            type: Hash
+    field :c, as: :display_collection, type: String
+
+    validates :date, presence: true
+    validates :metric, presence: true
+    validates :metric, uniqueness: { scope: %i[date display_collection] }
+
+    def self.spawn
+      return unless SupplejackApi.config.log_metrics == true
+      enable_metrics_logger
+
+      display_collections = find_display_collections_to_process
+      dates = find_dates_to_process
+
+      metrics = []
+
+      dates.each do |date|
+        METRICS_LOGGER.info "Current Date: #{date}"
+
+        display_collections.each do |dc|
+          METRICS_LOGGER.info "Current Display Collection: #{dc}"
+
+          METRICS.each do |metric|
+            METRICS_LOGGER.info "Current Metric: #{metric}"
+
+            record_metrics = record_metrics_to_be_processed(date, metric, dc)
+
+            results = calculate_results(record_metrics, metric)
+
+            # If there are no results for a metric, date, and display collection
+            # Skip to the next metric
+            next if results.select { |_key, value| value.positive? }.empty?
+
+            top_collection_metric = find_or_create_top_collection_metric(date, metric, dc)
+            update_top_collection_metric(top_collection_metric, results)
+            metrics.push(top_collection_metric)
+          end
+        end
+        stamp_record_metrics(date)
+      end
+
+      disable_metrics_logger
+
+      metrics
+    end
+
+    def self.find_dates_to_process
+      METRICS_LOGGER.info 'Collecting Dates.'
+
+      query = { :date.lt => Time.zone.now.beginning_of_day }
+      SupplejackApi::RecordMetric.where(query)
+                                 .map(&:date).uniq
+    end
+
+    def self.find_display_collections_to_process
+      METRICS_LOGGER.info 'Collecting Collections.'
+
+      query = { :date.lt => Time.zone.now.beginning_of_day, processed_by_top_collection_metrics: false }
+      SupplejackApi::RecordMetric.where(query)
+                                 .map(&:display_collection).uniq.compact
```

Comment:
Are you getting back nil display_collections?

Reply:
no we shouldn't. good spot - removed.